move_base/src/move_base.cpp

```diff
@@ -560,6 +565,90 @@ namespace move_base {
     return global_pose;
   }
 
+  //publish feedback based on latest_plan.Called in context of executeCycle.
+  void MoveBase::publishFeedback(const geometry_msgs::PoseStamped& current_position)
+  {
+    //calculate distance to goal from current position
+    std::pair<bool, double> distance_info = calculateDistanceToGoal(current_position);
+    if(!distance_info.first)
+    {
+      return;
+    }
+
+    //calculate average velocity
+    ros::Time current_time = ros::Time::now();
+    double dist = distance_info.second;
+    double average_vel = calculateAverageVelocity(dist, current_time);
+
+    //publish feedback
+    move_base_msgs::Feedback msg;
+    msg.dist_to_goal = dist;
+    msg.time_to_goal = prev_feedback_info_.getTime(dist, average_vel);
+    feedback_distance_pub_.publish(msg);
+
+    //save previous feedback info
+    prev_feedback_info_.time = current_time;
+    prev_feedback_info_.velocity = average_vel;
+    prev_feedback_info_.dist_to_goal = dist;
+  }
+
+  ///calculate distance from closest pose to current position to the goal. 
+  //return pair.The first parameter is status(true/false).The second is a distance.
+  std::pair<bool, double> MoveBase::calculateDistanceToGoal(const geometry_msgs::PoseStamped& current_position)
+  {
+    boost::unique_lock<boost::recursive_mutex> lock(planner_mutex_);
+    if(latest_plan_->empty())
+    {
+      return std::make_pair(false, 0.0);
+    }
+    std::vector<geometry_msgs::PoseStamped> plan;
+    plan = *latest_plan_;
+    lock.unlock();
+    //find the closest pose to current position, based on straigth line distance.
+    uint32_t min_index = 0;
+    double min_distance = distance(plan[0], current_position);
+    
+    for(uint32_t i = 1; i < plan.size(); ++i)
+    {
+      double current_distance = distance(plan[i], current_position);
+      if(current_distance < min_distance)
+      {
+        min_distance = current_distance;
+        min_index = i;
+      }
+    }
+    //calculate distance from closest pose to current position to the goal. 
+    double dist = 0.0;
+    for(uint32_t idx = min_index; idx < plan.size() - 1; ++idx)
+    {
+      dist += distance(plan[idx], plan[idx + 1]);
+    }
+
+    return std::make_pair(true, dist);
+  }
+
+  //the calculation based on "exponentially weighted moving average" (EWMA).
+  //average velocity = factor * previous average velocity + (1-factor) * current velocity
+  //factor between (0,1)
+  double MoveBase::calculateAverageVelocity(double dist_to_goal, const ros::Time& current_time)
+  {
+    double average_vel = FeedbackInfo::MIN_VELOCITY;
+
+    if(prev_feedback_info_.is_set)
+    {
+      ros::Duration dtime = current_time - prev_feedback_info_.time;
+      double current_velocity = (prev_feedback_info_.dist_to_goal - dist_to_goal)/dtime.toSec();
+      average_vel = FeedbackInfo::WEIGHT_AVERAGE_VELOCITY_FACTOR * prev_feedback_info_.velocity +
```

Comment:
Bias correction is needed in order to spin-up better.
I gave you wrong link to bias correction. Here is the right one https://www.youtube.com/watch?v=lWzo8CajF5s
Otherwise you depend too much on the 1st speed measurement, which can be noisy.